File: src/types/agent.ts
/**
 * @module types/agent
 * @description TypeScript type definitions for agents, workflows, and agent library
 *
 * PURPOSE:
 * - Define Agent interface for reusable Claude Code agent configurations
 * - Define LibraryAgent for pre-built agent catalog
 * - Define AgentTier for basic vs advanced complexity
 * - Define workflow and tool types for advanced agents
 *
 * EXPORTS:
 * - AgentTier - "basic" | "advanced" complexity level
 * - AgentCategory - Categories for organizing agents in the library
 * - AgentTool - Tool definition for advanced agents
 * - AgentWorkflowStep - Workflow step for advanced agents
 * - LibraryAgent - A pre-defined agent from the agent library catalog
 * - Agent - A saved agent with database fields
 * - AgentCategoryInfo - Metadata about an agent category (label, description, icon)
 *
 * PATTERNS:
 * - Types mirror Rust structs in models/agent.rs
 * - Agents have usage analytics (usageCount)
 * - Basic agents have instructions only, advanced have workflow + tools
 * - LibraryAgent has tags for tech-stack-based relevance scoring
 *
 * CLAUDE NOTES:
 * - Keep in sync with Rust models in src-tauri/src/models/agent.rs
 * - DateTime fields are serialized as ISO strings by Tauri
 * - TechTag is imported from skill.ts for consistency
 * - Advanced agents have workflow steps, tools, and trigger patterns
 */

import type { TechTag } from "./skill";

export type AgentTier = "basic" | "advanced";

export type AgentCategory =
  | "testing"
  | "code-review"
  | "documentation"
  | "debugging"
  | "refactoring"
  | "feature-development";

export interface AgentTool {
  name: string;
  description: string;
  required: boolean;
}

export interface AgentWorkflowStep {
  step: number;
  action: string;
  description: string;
}

export interface LibraryAgent {
  slug: string;
  name: string;
  description: string;
  tier: AgentTier;
  category: AgentCategory;
  tags: TechTag[];

  // Basic tier: instruction content
  instructions: string;

  // Advanced tier: workflow definition
  workflow?: AgentWorkflowStep[];
  tools?: AgentTool[];
  triggerPatterns?: string[];
}

export interface Agent {
  id: string;
  name: string;
  description: string;
  tier: AgentTier;
  category: AgentCategory;
  instructions: string;
  workflow: AgentWorkflowStep[] | null;
  tools: AgentTool[] | null;
  triggerPatterns: string[] | null;
  projectId: string | null;
  usageCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface AgentCategoryInfo {
  id: AgentCategory;
  label: string;
  description: string;
  icon: string;
}

File: src-tauri/src/models/agent.rs
//! @module models/agent
//! @description Data models for agents and workflows
//!
//! PURPOSE:
//! - Define Agent struct for reusable Claude Code agent configurations
//! - Define WorkflowStep for advanced agent workflows
//! - Define AgentTool for agent tool definitions
//!
//! DEPENDENCIES:
//! - serde - Serialization for Tauri IPC
//! - chrono - Timestamp handling
//!
//! EXPORTS:
//! - Agent - A reusable Claude Code agent configuration
//! - WorkflowStep - A step in an advanced agent workflow
//! - AgentTool - A tool definition for advanced agents
//!
//! PATTERNS:
//! - Agents have markdown instructions and optional workflow definitions
//! - Advanced agents have workflow steps, tools, and trigger patterns
//! - JSON fields (workflow, tools, trigger_patterns) are serialized/deserialized
//!
//! CLAUDE NOTES:
//! - Keep in sync with TypeScript types in src/types/agent.ts
//! - workflow, tools, trigger_patterns are Option<Vec<T>> for basic agents
//! - tier is "basic" or "advanced"

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Agent {
    pub id: String,
    pub name: String,
    pub description: String,
    pub tier: String,
    pub category: String,
    pub instructions: String,
    pub workflow: Option<Vec<WorkflowStep>>,
    pub tools: Option<Vec<AgentTool>>,
    pub trigger_patterns: Option<Vec<String>>,
    pub project_id: Option<String>,
    pub usage_count: u32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowStep {
    pub step: u32,
    pub action: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentTool {
    pub name: String,
    pub description: String,
    pub required: bool,
}
